xilem_core/src/message.rs

```diff
@@ -1,74 +1,99 @@
-                Message {
-                    id_path,
-                    body: Box::new(event),
-                }
-            }
-        }
-    };
-}
+use alloc::boxed::Box;
 
-/// A result wrapper type for event handlers.
+/// The possible outcomes from a [`View::message`]
+///
+/// [`View::message`]: crate::View::message
 #[derive(Default)]
-pub enum MessageResult<A> {
-    /// The event handler was invoked and returned an action.
+pub enum MessageResult<Action> {
+    /// An action for a parent message handler to use
     ///
-    /// Use this return type if your widgets should respond to events by passing
-    /// a value up the tree, rather than changing their internal state.
-    Action(A),
-    /// The event handler received a change request that requests a rebuild.
-    ///
-    /// Note: A rebuild will always occur if there was a state change. This return
-    /// type can be used to indicate that a full rebuild is necessary even if the
-    /// state remained the same. It is expected that this type won't be used very
-    /// often.
-    #[allow(unused)]
+    /// This allows for sub-sections of your app to use an elm-like architecture
+    Action(Action),
+    /// This event had no impact on the app state, or the impact it did have
+    /// does not require the element tree to be recreated.
     RequestRebuild,
-    /// The event handler discarded the event.
-    ///
-    /// This is the variant that you **almost always want** when you're not returning
-    /// an action.
-    #[allow(unused)]
     #[default]
+    /// This event had no impact on the app state, or the impact it did have
+    /// does not require the element tree to be recreated.
     Nop,
-    /// The event was addressed to an id path no longer in the tree.
-    ///
-    /// This is a normal outcome for async operation when the tree is changing
-    /// dynamically, but otherwise indicates a logic error.
-    Stale(Box<dyn Any>),
+    /// The view this message was being routed to no longer exists.
+    Stale(DynMessage),
 }
 
-// TODO: does this belong in core?
-pub struct AsyncWake;
+/// A dynamically typed message for the [`View`] trait.
+///
+/// Mostly equivalent to `Box<dyn Any>`, but with support for debug printing.
+///
+/// These messages must also be [`Send`].
+/// This makes using this message type in a multithreaded context easier.
+/// If this requirement is causing you issues, feel free to open an issue
+/// to discuss.
+/// We are aware of potential backwards-compatible workarounds, but
+/// are not aware of any tangible need for this.
+///
+/// [`View`]: crate::View
+pub type DynMessage = Box<dyn Message>;
+/// Types which can be contained in a [`DynMessage`].
+// The `View` trait could have been made generic over the message type,
+// primarily to enable flexibility around Send/Sync and avoid the need
+// for allocation.
+pub trait Message: 'static + Send {
```

Comment:
I'm currently running into the issue while rewriting xilem_web event listeners, that this requires `Send`. `web_sys::Event` doesn't implement `Send`.

So the `Send` bound should be optional (another reason to make it generic?).